### PR TITLE
WIP: [Breaking refactor] Converting DanfoJs to typescript

### PR DESCRIPTION
**Pull request by risenW**
This refactor converts the danfojs to use typescript. This is scheduled to be released as Danfo.Js v1.0 (stable), and will include a couple of breaking changes, bug fixes and new features. 

Task and progress
* [x] Generic core
* [x] Frame class
* [x] Series class 
* [x] File Readers /input and output 
* [x] Utility class
* [x] Configuration class
* [x] Groupby class
* [x] Concat
* [x] Merge class
* [x] Preprocessing class
* [x] Str class (String)
* [x] Dt class (Datetime)
* [x] Indexing
* [x] Plotting class 
* [x] Date range

## Bug Fixes
* [x] Column data not being updated when mutating internal data array
* [x] Str class issue for non-strings
* [x] Better error message
* [x] Fix support for all JS date string
* [x] Fix loc slicing bug for row index with string labels
* [x] Fix wrong order of axis used during computation. `axis=1` ==> row-wise operations and `axis=0` ==> column-wise operation
* [x] apply function now works only across a specified axis

## New Features
* [x] Ability to create an empty NDframe
* [x] Flag for toggling between low/high memory mode
* [x] `Inplace` support for all mutating operations
* [x] Ability to set Configuration values on NDframe creation
* [x] Support boolean mask for subsetting with iloc. E.g `df.iloc({rows: df["count"].gt(5), columns: [0, 1]})`
* [x] Support boolean mask for subsetting with loc. E.g `df.loc({rows: df["count"].gt(5), columns: ["Count", "Size"]})`
* [x] Update an existing column value via subsetting. E.g `df["count"] = [1,3,4,5]`
* [x] Add loc indexing support for Series
* [x] Add configuration support for formating table display in console
* [x] applyMap ==> Element wise apply function for DataFrame
* [X] `and` and `or` logical comparison support. E.g ` df.loc({rows: df['Salary_in_1000'].gte(100)).and(df['Age'].gt(60)) })`
* [x] Streaming support for CSV and JSON files into NDframes
* [x] New IO function
   * [x] streamJSON ==> Supports streaming of local or remote JSON files into DataFrame.
   * [x] streamCSV ==> Supports streaming of local or remote CSV files into DataFrame.
   * [x] openCsvInputStream ==> Open a local/remote CSV file as a readable stream
   * [x] writeCsvOutputStream Open a local/remote CSV file as a writable stream
* [x] readCSV supports config values for headers, separator, etc. 
* [x] toCSV supports config values for output 
* [x] readJSON supports config values for headers, separator, etc. 
* [x] toJSON supports config values for formating output 
* [x] Query/Filter with multiple condition support. E.g ` df.loc({rows: df['Salary_in_1000'].gte(100)).and(df['Age'].gt(60)) })`
* [x] streamCsvTransforme ==> Pipable stream transformer for incrementally transforming DataFrames

### CODE REVIEW

**Comment by JhennerTigreros** [comment]
Hi @risenW, which is the current process on this? I think we can divide the migration to TS by chunks in the current codebase 😄

**Comment by risenW** [comment]
> Hi @risenW, which is the current process on this? I think we can divide the migration to TS by chunks in the current codebase 😄

Thanks for offering to help out here @JhennerTigreros So just give me a few more days to complete the Generic core class. Once done you can start working on the other classes. I'm doing some major rewrites in the way NDFrame data is created and stored, to solve some memory constraints in the old version.  

I'll update this issue when I'm done, and then you can start contributing here.

**Comment by lilred** [comment]
@risenW out of curiosity, will data frames' types reflect their columns? TypeScript is one of the rare languages whose type system is sufficiently advanced to e.g. faithfully represent a SQL JOIN operation (`merge` in Pandas parlance). Is this part of the plan?

**Comment by risenW** [comment]
> @risenW out of curiosity, will data frames' types reflect their columns? TypeScript is one of the rare languages whose type system is sufficiently advanced to e.g. faithfully represent a SQL JOIN operation (`merge` in Pandas parlance). Is this part of the plan?

@lilred I didn’t get your question. Can you rephrase?

**Comment by lilred** [comment]
Disclaimer: I'm not acquainted with Danfo yet.

A data frame could have type `DataFrame<T>` where `T` is an object describing the column names and their types. a value of type `DataFrame<{a: number, b: bool}>` is a data frame with two columns, one (`a`) number-valued and one (`b`) boolean-valued.

I believe it's possible in TypeScript to define a SQL JOIN-like operation such that if `df1` has type `DataFrame<{a: number, b: bool}>` and `df2` has type` DataFrame<{a: number, c: string}>` then `df1.join(df2, "outer", ["a"])` has type `DataFrame<{a: number, b?: bool, c?: string}>`, *mutatis mutandis* for inner/left/right joins. Then data frame access can be performed without runtime type checks or column index checks.

**Comment by JhennerTigreros** [comment]
Ohh that's sound really good, I will wait until you finish the migration and start porting the correlation functions to TypeScript and tackle performance issues. I think we can use web assembly to certain heavily operations. What do you think?

**Comment by risenW** [comment]
> Ohh that's sound really good, I will wait until you finish the migration and start porting the correlation functions to TypeScript and tackle performance issues. I think we can use web assembly to certain heavily operations. What do you think?

Yes, that would help a lot, especially for areas where we are not using tensorflow. 
cc @steveoni 

Disclaimer: I do not know web assembly-yet 😊

**Comment by pguso** [comment]
Really nice that there are coming so many new features!

Is there an easy way to convert one column of a DataFrame to a Series object already?

Will there be pandas features like this `df.columns.str.upper().str.replace('_', '')` or making date columns easily to the index?

**Comment by risenW** [comment]
> Really nice that there are coming so many new features!
> 
> Is there an easy way to convert one column of a DataFrame to a Series object already?
> 
> Will there be pandas features like this `df.columns.str.upper().str.replace('_', '')` or this `df.index.str.replace(".*(?=Test)", "")` or making date columns easily to the index?

Yes, referencing a column using subseting like df[“col1”] will return a Series. 

A call to df.columns or df.index will return an array of the values. So you can map that to any function.

**Comment by risenW** [comment]
Closing this PR. It has been merged and released in the latest v1.0.0
